[Fix #4508] Added return nil cop

donjar opened the pull request:
This new cop checks for 'return nil' or 'return' by default, and changes it to one of them based on configuration.

Tests are still failing though:
```
     Failure/Error: cop_config[supported_styles].map(&:to_sym)
     
     NoMethodError:
       undefined method `map' for nil:NilClass
     # ./lib/rubocop/cop/mixin/configurable_enforced_style.rb:88:in `supported_styles'
     # ./lib/rubocop/cop/mixin/configurable_enforced_style.rb:66:in `style'
     # ./lib/rubocop/cop/style/return_nil.rb:48:in `correct_style?'
     # ./lib/rubocop/cop/style/return_nil.rb:31:in `on_return'
     # ./lib/rubocop/cop/commissioner.rb:44:in `block (2 levels) in on_return'
     # ./lib/rubocop/cop/commissioner.rb:111:in `with_cop_error_handling'
     # ./lib/rubocop/cop/commissioner.rb:43:in `block in on_return'
     # ./lib/rubocop/cop/commissioner.rb:42:in `each'
     # ./lib/rubocop/cop/commissioner.rb:42:in `on_return'
     # ./lib/rubocop/ast/traversal.rb:12:in `walk'
     # ./lib/rubocop/cop/commissioner.rb:60:in `investigate'
     # ./lib/rubocop/rspec/cop_helper.rb:75:in `_investigate'
     # ./lib/rubocop/rspec/cop_helper.rb:51:in `autocorrect_source'
     # ./spec/rubocop/cop/style/return_nil_spec.rb:48:in `block (3 levels) in <top (required)>'
```

I thought it must be something related to the configuration of the cops in the specs... I'm not sure, can anyone help? I'll check again tomorrow if I have time.

Shoutout to @Drenmi - I was there in the Singapore Ruby meetup a few weeks back and he presented about Rubocop (one of my favorite gems, by the way). Was seriously considering to contribute to open source, and his talk made the push for me :) thanks Ted!

-----------------

Before submitting the PR make sure the following are checked:

* [X] Wrote [good commit messages][1].
* [X] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [X] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [X] Added tests.
* [X] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* [X] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [X] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [X] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

Drenmi commented on config/default.yml:
```diff
@@ -1154,6 +1154,12 @@ Style/RegexpLiteral:
+Style/ReturnNil:
+  EnforcedStyle: return
+  SupportedStyles:
+    - return # return
```
These comments aren't needed, since they're identical to the actual options. 🙂

Drenmi commented on lib/rubocop/cop/style/return_nil.rb:
```diff
@@ -0,0 +1,53 @@
+      # Supported styles are: return, return_nil.
+      #
+      # @example
+      #   # bad if EnforcedStyle is return, good if EnforcedStyle is return_nil
```
Please add a section for each `EnforcedStyle`, as it's easier to understand. 🙂 Also, the example right now would cause another RuboCop offense (redundant return.) It would be nice if the examples were non-offending code. Metasyntactic variables are a conventional way to indicate that a name doesn't matter, i.e. `abcd` -> `foo`.

Grouping example:

```
# @example
#
#   # EnforcedStyle: return (default)
#
#   # bad
#   def foo
#     return nil
#   end
#
#   # good
#   def foo
#     return
#   end
```

Drenmi commented on lib/rubocop/cop/style/return_nil.rb:
```diff
@@ -0,0 +1,53 @@
+        end
+
+        def correct_style?(node)
+          ((style == :return) && !return_nil_node?(node)) ||
```
Not sure why the negation is here? I find `style == :return && return_node?(node)` quite straightforward. 🙂 

Also, all the parens are redundant. 😉

Drenmi commented on spec/rubocop/cop/style/return_nil_spec.rb:
```diff
@@ -0,0 +1,55 @@
+  context 'when enforced style is `return`' do
+    let(:config) do
+      RuboCop::Config.new(
+        'Style/ReturnNil' => {
```
I think the specs are failing because this doesn't contain the supported styles.

Drenmi commented:
Great work! I left some comments. Have a look. 🙂 

You'll need to fix the failing CI and rebase to fix conflicts in `CHANGELOG`. 🙂

donjar commented on lib/rubocop/cop/style/return_nil.rb:
```diff
@@ -0,0 +1,68 @@
+        RETURN_MSG = 'Use `return` instead of `return nil`.'.freeze
+        RETURN_NIL_MSG = 'Use `return nil` instead of `return`.'.freeze
+
+        def_node_matcher :return_nil_node?, '(return (nil))'
```
Hey @Drenmi - there must be something wrong here from what I see. What should be the correct node matcher for `return nil`?

I tried using `ruby-parse`:

```
$ ruby-parse -e 'return nil'
(return
  (nil))
```

Drenmi replied:
Yes. `(return (nil))` looks right. Because `nil` literal is parsed as its own node. 🙂

donjar replied:
Yep, but it causes an error when using Rubocop somehow... You can check Travis, it fails (https://travis-ci.org/bbatsov/rubocop/jobs/261506679). I'm still not sure what happened :/

donjar replied:
@Drenmi ping :)

Drenmi replied:
@donjar: Will have time to look at this tomorrow. Sorry for the delay. 🙂

donjar commented:
@Drenmi ping :)

bbatsov commented:
This should be squashed and rebased.

donjar commented:
I still cannot figure out why the specs failed >.<

Drenmi commented:
@donjar: I'm on holiday right now, but can help check next week. 🙂

bbatsov commented:
Seems like an unhandled `nil` somewhere in your code.

Drenmi commented:
@donjar: Looks like a problem in `NodePattern` with handling `nil` literals. In this case, `(return nil)` should match, but it doesn't.

You can work around it for now by using a regular method for `#return_nil_node?`:

```ruby
def return_nil_node?(node)
  !node.children.empty? && node.children.first.nil_type?
end
```

I will have a look at the `NodePattern` issue as soon as I have time.

donjar commented:
@Drenmi that works! I've added a `TODO` comment as well.

bbatsov commented:
Seems that the build is still failing, though.

Drenmi commented:
@donjar: Try re-running `rake generate_cops_documentation`, commit and squash the changes, and push again.

donjar commented:
It should be working now - I've added a few specs as well.

I've changed some of the codebase though:
- There are actually some parts that does `return nil` instead of `return` in the codebase, so I've changed them.
- I'm still using `return nil` instead of `return` inside blocks, to prevent `Lint/NonLocalExitFromIterator` offenses.

Drenmi commented:
> There are actually some parts that does return nil instead of return in the codebase, so I've changed them.

This is correct. RuboCop should pass when run on itself. 🙂 

> I'm still using return nil instead of return inside blocks, to prevent Lint/NonLocalExitFromIterator offenses.

This sounds like a bug in `Lint/NonLocalExitFromIterator` to me. 🤔

bbatsov commented:
> This is correct. RuboCop should pass when run on itself. 🙂

Or we should disable the cop by default. For me there's a difference between `return` and `return nil` semantically - I use the first just to halt evaluation and the second if I really care about the return value. Although it's generally a bad idea to return `nil` anyways.

donjar commented:
@bbatsov @Drenmi so I have made this cop be disabled by default and reverted all changes to other files. However, I still retained the fix related to `Lint/NonLocalExitFromIterator`.

Drenmi commented:
> For me there's a difference between return and return nil semantically - I use the first just to halt evaluation and the second if I really care about the return value. Although it's generally a bad idea to return nil anyways.

I very rarely write methods that return `nil`, but I definitely agree that there's a semantic difference, and it's a good way to indicate whether the return value is intended to be used or not. 🙂
